File: 02_Object_Oriented_Programming/02_Class_Relationships/05_Dependency/python/main.py
"""
Driver — demonstrates Dependency (USES-A temporarily).
Run: python main.py
"""

from document import Document
from printer import Printer


def main() -> None:
    doc = Document("Report", "Hello, World!")
    office_printer = Printer("Office HP")
    home_printer = Printer("Home Canon")

    # Same document, different printers — dependency is temporary
    doc.print_document(office_printer)
    print()
    doc.print_document(home_printer)


if __name__ == "__main__":
    main()

"""
Expected Output:
Document: Report
[Office HP] Printing: Hello, World!

Document: Report
[Home Canon] Printing: Hello, World!
"""

File: 02_Object_Oriented_Programming/02_Class_Relationships/05_Dependency/javascript/main.js
/**
 * Driver — demonstrates Dependency (USES-A temporarily).
 * Run: node main.js
 */

const Document = require("./document");
const Printer = require("./printer");

const doc = new Document("Report", "Hello, World!");
const officePrinter = new Printer("Office HP");
const homePrinter = new Printer("Home Canon");

// Same document, different printers — dependency is temporary
doc.printDocument(officePrinter);
console.log();
doc.printDocument(homePrinter);

/*
Expected Output:
Document: Report
[Office HP] Printing: Hello, World!

Document: Report
[Home Canon] Printing: Hello, World!
*/
